rcl_action/include/rcl_action/action_server.h

```diff
@@ -0,0 +1,667 @@
+ * remain constant during publish.
+ *
+ * This function is thread safe so long as access to both the action server and
+ * `ros_feedback` is synchronized.
+ * That means that calling rcl_action_publish_feedback() from multiple threads
+ * is allowed, but calling rcl_action_publish_feedback() at the same time as
+ * non-thread safe action server functions is not, e.g. calling
+ * rcl_action_publish_feedback() and rcl_action_server_fini() concurrently is not
+ * allowed.
+ * Before calling rcl_action_publish_feedback() the message can change and after
+ * calling rcl_action_publish_feedback() the message can change, but it cannot be
+ * changed during the publish call.
+ * The same `ros_feedback`, however, can be passed to multiple calls of
+ * rcl_action_publish_feedback() simultaneously, even if the action servers differ.
+ * `ros_feedback` is unmodified by rcl_action_publish_feedback().
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | Yes [1]
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ * <i>[1] for unique pairs of action servers and responses, see above for more</i>
+ *
+ * \param[in] action_server handle to the action server that will publish the feedback
+ * \param[in] goal_handle a struct with the goal handle associated with the feedback
+ * \param[in] ros_feedback a struct containing the goal feedback message
+ * \return `RCL_RET_OK` if the response was sent successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs. *
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_publish_feedback(
+  const rcl_action_server_t * action_server,
+  const rcl_action_goal_handle_t * goal_handle,
+  void * ros_feedback);
+
+/// Publish a status array message for accepted goals associated with a rcl_action_server_t.
+/**
+ *
+ * This function acts like a ROS publisher and is potntially a blocking call.
+ * \see rcl_publish()
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] action_server handle to the action server that will publish the status message
+ * \param[in] status_message a struct with the status array to publish
+ * \return `RCL_RET_OK` if the response was sent successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_publish_status(
+  const rcl_action_server_t * action_server,
+  const rcl_action_status_array_t * status_mesage);
```

Comment:
I would have expected the `action_server` to internally know the status of each goal, and if so it wouldn't need to be passed in here. Where is the state for each goal tracked?

Reply:
I think the action server will track the state of each goal. I figured it might provide some more flexibility to separate the publish status and the getter for the array of goal handles. We can add an additional publish status function that does not take a status message for convenience.